### PR TITLE
Support installing packages under custom name as a way of avoiding conflicts

### PR DESCRIPTION
**Pull request by kasperski95**
`python protostar.py install ... --name foo`

### CODE REVIEW

File: src/commands/install/install_package_from_repo.py [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+    destination: str,
+):
+    logger = getLogger()
+
+    try:
+        repo = Repo(repo_root_dir)
+    except InvalidGitRepositoryError as _err:
+        raise installation_exceptions.InvalidLocalRepository()
+    except NoSuchPathError as _err:
+        raise installation_exceptions.InvalidLocalRepository()
```
Comment (MaksymilianDemitraszek) [inline src/commands/install/install_package_from_repo.py:27]:
```suggestion
```
Why do we catch this exceptions?

File: src/commands/install/install_package_from_repo_test.py [hunk 2]
```diff
@@ -0,0 +1,27 @@
...
+import pytest
+from git.repo import Repo
+
+from src.commands.install import installation_exceptions
+from src.commands.install.install_package_from_repo import install_package_from_repo
+
+
+@pytest.fixture(name="repo_url")
+def fixture_repo_url():
+    return "https://github.com/starkware-libs/cairo-lang"
```
Comment (MaksymilianDemitraszek) [inline src/commands/install/install_package_from_repo_test.py:12]: I think we should't call external source from our tests, it is more problematic than it brings value.

File: src/commands/install/extract_info_from_repo_id.py [hunk 3]
```diff
@@ -55,16 +62,16 @@ def extract_info_from_repo_id(repo_id: str) -> PackageInfo:
...
-def _map_ssh_to_url(ssh: str) -> Optional[str]:
+def _map_ssh_to_url(ssh: str) -> str:
     slug = _extract_slug_from_ssh(ssh)
     domain_match = re.search(r"(?<=git@).*(?=:)", ssh)
 
     if domain_match is None:
-        return None
+        raise installation_exceptions.InvalidPackageName("Couldn't map SSH to URL.")
 
     return f"https://{domain_match.group()}/{slug}"
```
Comment (MaksymilianDemitraszek) [inline src/commands/install/extract_info_from_repo_id.py:76]: What do you think about writing tests for those functions? It can be error prone

**Review by MaksymilianDemitraszek — COMMENTED** [review]
Generally very good job, some ideas/questions in the comments

File: src/commands/install/handle_install_command.py [hunk 4]
```diff
@@ -0,0 +1,20 @@
+from os import getcwd
+from typing import Any
+
+from src.commands.install.extract_info_from_repo_id import extract_info_from_repo_id
+from src.commands.install.install_package_from_repo import install_package_from_repo
+
+
+def handle_install_command(args: Any) -> None:
+    if args.command != "install":
```
[the remark this answers is not in the document]
Comment (kasperski95) [reply src/commands/install/handle_install_command.py:9]: Nope. The function skips installation process if a command is different from `install`. (There's `return` below that if.)

File: src/commands/install/handle_install_command.py [hunk 5]
```diff
@@ -0,0 +1,20 @@
+from os import getcwd
+from typing import Any
+
+from src.commands.install.extract_info_from_repo_id import extract_info_from_repo_id
+from src.commands.install.install_package_from_repo import install_package_from_repo
+
+
+def handle_install_command(args: Any) -> None:
```
[the remark this answers is not in the document]
Comment (kasperski95) [reply src/commands/install/handle_install_command.py:8]: - `install` seems to be too generic name in hindsight. Personally, I prioritize readability over consistency.
- `Namespace` type doesn't provide correct typing. `Any` type informs developer to not rely on typing.

File: src/commands/install/install_package_from_repo.py [hunk 6]
```diff
@@ -0,0 +1,49 @@
...
+from git.objects import Submodule
+from git.repo import Repo
+
+from src.commands.install import installation_exceptions
+
+
+def install_package_from_repo(
+    name: str,
+    url: str,
+    tag: Optional[str],
```
[the remark this answers is not in the document]
Comment (kasperski95) [reply src/commands/install/install_package_from_repo.py:16]: Ok

File: src/commands/install/install_package_from_repo.py (hunk 1 above)
Comment (kasperski95) [reply src/commands/install/install_package_from_repo.py:27]: It is a foundation for making errors “user-friendly”.  I am planning to create an abstract `ProtostarException` that all custom exceptions in the app should extend. In this way, we can later determine whether to display a nice error message, e.g. “Git repository is not initialized. Please run `git init`.” or  yeet stack trace.

File: src/commands/install/install_package_from_repo_test.py (hunk 2 above)
Comment (kasperski95) [reply src/commands/install/install_package_from_repo_test.py:12]: Agree. I want lazy here.

File: src/commands/install/extract_info_from_repo_id.py (hunk 3 above)
Comment (kasperski95) [reply src/commands/install/extract_info_from_repo_id.py:76]: Ok

File: src/commands/install/install_package_from_repo.py (hunk 1 above)
Comment (kasperski95) [reply src/commands/install/install_package_from_repo.py:27]: I'll add handling validation error in this task: https://github.com/software-mansion/protostar/issues/33.

File: protostar.py [hunk 7]
```diff
@@ -33,6 +32,9 @@ def cairo_file(file_path: str) -> Path:
 
 cmd_install_parser = root_subparsers.add_parser("install")
 cmd_install_parser.add_argument("package", type=str)
+cmd_install_parser.add_argument(
+    "--name", type=str, help="Custom package's directory name"
```
Comment (Arcticae) [inline protostar.py:36]: is that argument required?

File: src/commands/install/handle_install_command.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (Arcticae) [reply src/commands/install/handle_install_command.py:9]: Actually we use this function like this:
```
   if args.command == "install":
        handle_install_command(args)
```
So maybe assertion would be more appropriate?

**Review by Arcticae — CHANGES_REQUESTED** [review]

File: protostar.py (hunk 7 above)
Comment (kasperski95) [reply protostar.py:36]: It's optional. All named arguments are optional by default.

File: src/commands/install/handle_install_command.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (kasperski95) [reply src/commands/install/handle_install_command.py]: Added assertion.

**Review by Arcticae — APPROVED** [review]

**Pull request merged**